### PR TITLE
Check that threads calling PThread.getTLS are bound.

### PR DESCRIPTION
**Pull request by facundominguez**
cc @mboes @robinbb

I took `cbits/helpers.c` from the latest hackage release.

### CODE REVIEW

**Comment by facundominguez** [comment]
cc @robinbb

**Comment by facundominguez** [comment]
I don't remember why this PR was closed. Possibly I intended to resubmit it after other changes. In any case, here it is rebased.

File: tests/Main.hs [hunk 1]
```diff
@@ -33,7 +34,7 @@ testIt name mkTLS getTLS allTLS freeAllTLS = do
   tls <- mkTLS (do putStrLn "  New() called.."
                    newIORef (-1 :: Int))
   mvs <- sequence $ replicate numCap newEmptyMVar
-  forM_ (zip [0..] mvs) $ \(ix,mv) -> forkOn ix $ do
+  forM_ (zip [0..] mvs) $ \(ix,mv) -> forkOS $ do
```
Comment (RyanGlScott) [inline tests/Main.hs]: That's, uh, kind of a miracle that this was working before. I wonder if it just happened to keep the same OS threads per capability for this one program, and so the TLS managed to work out in the end?

File: Data/TLS/PThread/Internal.hs [hunk 2]
```diff
@@ -92,6 +100,13 @@ getTLS TLS{key,mknew,allCopies} = do
     return a
    else
     deRefStablePtr p
+ where
+  checkBoundness = when rtsSupportsBoundThreads $ do
+    b <- isCurrentThreadBound
+    when (not b) $ do
+      let msg = "thread-local-storage: PThread.getTLS used from an unbound thread."
+      hPutStrLn stderr msg
+      error msg
```
Comment (RyanGlScott) [inline Data/TLS/PThread/Internal.hs]: Do we need to both print the message to `stderr` and call `error`? Alternatively, couldn't we just call `fail`?

Comment (facundominguez) [inline Data/TLS/PThread/Internal.hs]: Addressed.

File: tests/Main.hs (hunk 1 above)
Comment (facundominguez) [inline tests/Main.hs]: It would make sense, but I don't know how the scheduler works.

**Comment by RyanGlScott** [comment]
Thanks!

**Pull request merged**